core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ImplementTableFunctionSource.java

```diff
@@ -265,18 +317,243 @@ private static NodeWithSymbols appendHelperSymbolsForSource(
+                .map(Symbol::toSymbolReference)
+                .collect(toImmutableList());
+        Expression rightRowNumber = right.getRowNumber().toSymbolReference();
+        Expression rightPartitionSize = right.getPartitionSize().toSymbolReference();
+        List<Expression> rightPartitionBy = right.getPartitionBy().stream()
+                .map(Symbol::toSymbolReference)
+                .collect(toImmutableList());
+
+        List<Expression> coPartitionConjuncts = Streams.zip(
+                        leftPartitionBy.stream(),
+                        rightPartitionBy.stream(),
+                        (leftColumn, rightColumn) -> new NotExpression(new ComparisonExpression(IS_DISTINCT_FROM, leftColumn, rightColumn)))
+                .collect(toImmutableList());
+        // In StatementAnalyzer we require that co-partitioned tables have non-empty partitioning column lists.
+        // Co-partitioning tables with empty partition by would be ineffective.
+        checkState(!coPartitionConjuncts.isEmpty(), "co-partitioned tables must have partitioning columns");
+
+        // Align matching partitions (co-partitions) from left and right source, according to row number.
+        // Matching partitions are identified by their corresponding partitioning columns being NOT DISTINCT from each other.
+        // If one or both sources are ordered, the row number reflects the ordering.
+        // The second and third disjunct in the join condition account for the situation when partitions have different sizes.
+        // It preserves the outstanding rows from the bigger partition, matching them to the first row from the smaller partition.
+        //
+        // (P1_1 IS NOT DISTINCT FROM P2_1) AND (P1_2 IS NOT DISTINCT FROM P2_2) AND ...
+        // AND (
+        //      R1 = R2
+        //      OR
+        //      (R1 > S2 AND R2 = 1)
+        //      OR
+        //      (R2 > S1 AND R1 = 1))
+        Expression joinCondition = new LogicalExpression(AND, Streams.concat(
+                        coPartitionConjuncts.stream(),
+                        Stream.of(new LogicalExpression(OR, ImmutableList.of(
+                                new ComparisonExpression(EQUAL, leftRowNumber, rightRowNumber),
+                                new LogicalExpression(AND, ImmutableList.of(
+                                        new ComparisonExpression(GREATER_THAN, leftRowNumber, rightPartitionSize),
+                                        new ComparisonExpression(EQUAL, rightRowNumber, new GenericLiteral("BIGINT", "1")))),
+                                new LogicalExpression(AND, ImmutableList.of(
+                                        new ComparisonExpression(GREATER_THAN, rightRowNumber, leftPartitionSize),
+                                        new ComparisonExpression(EQUAL, leftRowNumber, new GenericLiteral("BIGINT", "1"))))))))
+                .collect(toImmutableList()));
```

Comment:
The formatting here looks off. It's not clear what `.collect(...)` attaches to. Reformat as follows for clarity:

```java
Expression joinCondition = new LogicalExpression(
        AND,
        Streams.concat(
                        coPartitionConjuncts.stream(),
                        Stream.of(new LogicalExpression(OR, ImmutableList.of(
                                new ComparisonExpression(EQUAL, leftRowNumber, rightRowNumber),
                                new LogicalExpression(AND, ImmutableList.of(
                                        new ComparisonExpression(GREATER_THAN, leftRowNumber, rightPartitionSize),
                                        new ComparisonExpression(EQUAL, rightRowNumber, new GenericLiteral("BIGINT", "1")))),
                                new LogicalExpression(AND, ImmutableList.of(
                                        new ComparisonExpression(GREATER_THAN, rightRowNumber, leftPartitionSize),
                                        new ComparisonExpression(EQUAL, leftRowNumber, new GenericLiteral("BIGINT", "1"))))))))
                .collect(toImmutableList()));
```

also, it might be slightly easier to read with a list builder instead of creating synthetic streams for the purpose of concatenating them:

```java
Expression joinCondition = new LogicalExpression(
        AND,
        ImmutableList.<Expression>builder()
                .addAll(coPartitionConjuncts)
                .add(new LogicalExpression(OR, ImmutableList.of(
                        new ComparisonExpression(EQUAL, leftRowNumber, rightRowNumber),
                        new LogicalExpression(AND, ImmutableList.of(
                                new ComparisonExpression(GREATER_THAN, leftRowNumber, rightPartitionSize),
                                new ComparisonExpression(EQUAL, rightRowNumber, new GenericLiteral("BIGINT", "1")))),
                        new LogicalExpression(AND, ImmutableList.of(
                                new ComparisonExpression(GREATER_THAN, rightRowNumber, leftPartitionSize),
                                new ComparisonExpression(EQUAL, leftRowNumber, new GenericLiteral("BIGINT", "1")))))))
                .build());
```